### PR TITLE
Fixed: Generation error when multiple Edges hanging (not merged)

### PR DESCRIPTION
**Pull request by xescugc**
When multiple Edges Hanging we had an issue that they where left alone.

Also for big graphs with nested connections we where having an issue to calculate the shortest path to the next Node (`visited` was shared between different paths so it was incorrect).

### CODE REVIEW

File: generate/connection.go [hunk 1]
```diff
@@ -64,8 +64,17 @@ func findEdgeConnections(g *graph.Graph, n *graph.Node, visited map[string]struc
 			})
 		} else {
 
+			// We make a copy of the visited so
+			// each nested edge does not share the
+			// same copy of it and produces a wrong
+			// result
+			aux := make(map[string]struct{})
```
Comment (tormath1) [inline generate/connection.go:71]: can be init using the length of `visited` ?

File: generate/connection.go [hunk 2]
```diff
@@ -75,7 +84,14 @@ func findEdgeConnections(g *graph.Graph, n *graph.Node, visited map[string]struc
 					Direction: direc,
 				},
 			}
-			res = append(res, append(con, cons...))
+
+			// If there are no connections means it's an end Node,
+			// and as it is not a Node (it'll have entered the previous
+			// condition) it means it has no Node at the end so we do not
+			// return it
+			if len(cons) != 0 {
```
Comment (tormath1) [inline generate/connection.go:92]: we can remove the `if` conditions ? 

the append works even if `cons`'s length is 0

File: generate/state.go [hunk 3]
```diff
@@ -206,6 +206,28 @@ func cleanHangingEdges(g *graph.Graph, opt Options) error {
...
+	// are not meant to be displayed
+	if opt.Connections {
+	RESTART:
+		for _, n := range g.Nodes {
+			pv, rs, err := getProviderAndResource(n.Canonical, opt)
+			if err != nil {
+				return err
+			}
+			if pv.IsEdge(rs) {
+				err := g.RemoveNodeByID(n.ID)
```
Comment (tormath1) [inline generate/state.go:221]:
```suggestion
				if err := g.RemoveNodeByID(n.ID); err != nil
```

File: generate/state_test.go [hunk 4]
```diff
@@ -95,6 +95,48 @@ func TestFromState_AWS(t *testing.T) {
 		assertEqualGraph(t, eg, g, cfg)
 	})
 
+	t.Run("Issue", func(t *testing.T) {
+		src, err := ioutil.ReadFile("./testdata/aws_state_error.json")
```
[the remark this answers is not in the document]
Comment (xescugc) [reply generate/state_test.go:99]: It's is pruned but not with `--canonicals` so it's easy to read and test, also the names do not share any information.

I'll update the name of the test to be `PR33` as this is more of an edge case test.

File: graph/graph.go [hunk 5]
```diff
@@ -216,18 +218,26 @@ func (g *Graph) InvertEdge(eID string) {
...
+func (g *Graph) RemoveNodeByID(ID string) error {
+	var found bool
 	for i, n := range g.Nodes {
 		if n.ID == ID {
+			found = true
 			idx := i
 			g.removeNodeByIDX(idx)
 		}
 	}
+	if !found {
```
[the remark this answers is not in the document]
Comment (xescugc) [reply graph/graph.go:232]: Yep totally :)

File: generate/connection.go (hunk 2 above)
Comment (xescugc) [reply generate/connection.go:92]: Yeh but the logic is `res = append(res, append(con, cons...))` so it's to prevent to append the `con`.

I'll move the logic to be more clear and no initialize the `con` if not needed.

**Comment by xescugc** [comment]
Done and commented.

File: generate/state_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tormath1) [reply generate/state_test.go]: Ok, but we have two files: 
* aws_state_pr33.json
* aws_state_error.json 

for only one test, so the `aws_state_error.json` can be removed no ?

File: graph/graph.go [hunk 6]
```diff
@@ -216,18 +217,24 @@ func (g *Graph) InvertEdge(eID string) {
 	}
 }
 
-// removeNodeByID removes the Node with the ID
-func (g *Graph) removeNodeByID(ID string) {
+// RemoveNodeByID removes the Node with the ID and the Edges
+// associated with it
+func (g *Graph) RemoveNodeByID(ID string) error {
+	var found bool
```
Comment (tormath1) [inline graph/graph.go:223]: you can remove it now :)

**Comment by xescugc** [comment]
Done

File: graph/graph.go [hunk 7]
```diff
@@ -216,18 +217,22 @@ func (g *Graph) InvertEdge(eID string) {
 	}
 }
 
-// removeNodeByID removes the Node with the ID
-func (g *Graph) removeNodeByID(ID string) {
+// RemoveNodeByID removes the Node with the ID and the Edges
+// associated with it
+func (g *Graph) RemoveNodeByID(ID string) error {
```
Comment (xlr-8) [inline graph/graph.go:222]: This will now return an error an empty graph, is it expected? I don't think we have situation's like this but that's new

**Comment by tormath1** [comment]
looks good, you can rebase !

**Comment by xescugc** [comment]
I'll w8 for @xlr-8 to say it too.
